rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/InfrastructureManager.java

```diff
@@ -80,37 +84,103 @@
     /** class' logger */
     protected static final Logger logger = Logger.getLogger(InfrastructureManager.class);
 
+    /** delay that is waited at startup of the scheduler to initialize the
+     * variables and persist them
+     */
+    private static final int PERSISTED_INFRA_VARIABLES_INIT_DELAY = 2000;
+
     /** manager's node source */
     protected NodeSource nodeSource;
 
-    /** deploying nodes list */
-    private final Map<String, RMDeployingNode> deployingNodes = new ConcurrentHashMap<>();
+    /** key to retrieve the deploying nodes list: Map<String, RMDeployingNode> */
+    private static final String DEPLOYING_NODES_KEY = "infrastructureManagerDeployingNodes";
 
-    private final Map<String, RMDeployingNode> lostNodes = new ConcurrentHashMap<>();
+    /** key to retrieve the lost nodes list: Map<String, RMDeployingNode> */
+    private static final String LOST_NODES_KEY = "infrastructureManagerLostNodes";
 
     /**
-     * node list, miror of nodesource.getAliveNodes(), to implement random
+     * key to retrieve node list, miror of nodesource.getAliveNodes(), to implement random
      * access
      */
-    private Hashtable<String, Node> acquiredNodes = new Hashtable<>();
+    private static final String ACQUIRED_NODES_KEY = "infrastructureManagerAcquiredNodes";
+
+    private static final String USING_DEPLOYING_NODES_KEY = "usingDeployingNodes";
 
-    private final ReentrantLock nodeAcquisitionLock = new ReentrantLock();
+    private static final String NB_DOWN_NODES_KEY = "infrastructureManagerNbDownNodes";
 
-    // shared fields, needs to be volatile not to cache it
-    private volatile boolean usingDeployingNodes = false;
+    protected static final String RM_URL_KEY = "infrastructureManagerRmUrl";
 
-    // shutdown marker
-    private volatile boolean shutdown = false;
+    /**
+     * Indicates whether the infrastructure is shutting down.
+     */
+    private AtomicBoolean shutDown = new AtomicBoolean(false);
 
     // used to timeout the nodes
     private transient Timer timeouter = null;
 
-    protected String rmUrl;
+    /**
+     * Store information about the running infrastructure. The map holds the name of monitored information and its
+     * value. The variables stored in this map should allow the full recovery of an infrastructure state.
+     * All accesses to this map must be synchronized.
+     */
+    protected Map<String, Serializable> persistedInfraVariables = new HashMap<>();
```

Comment:
That would double protect the map because it is always accessed within a read or write lock acquisition through the `getPersistedInfraVariable` or `setPersistedInfraVariable` (the lock is needed because usually, when manipulating this variable, a sequence of operation must be atomic). But indeed, if we want to protect against future code that would use this variable without a lock, that would be a good move. What do you think after all?

Reply:
Okay I see. Yes the SynchronizedHashMap doesn not synchronise a series of operations, so it is of no use here. 👍